stdlib/dynarray.ml

```diff
@@ -0,0 +1,737 @@
+  else begin
+    a.length <- length_a + length_b;
+    (* END ATOMIC
+
+       Notice that, unlike for [add_last], the atomic section here
+       lasts until the length is extended, but stops before the
+       elements are added, so one could observe missing elements if
+       the code yields.  *)
+    (* Note: we intentionally update the length *before* filling the
+       elements. This "reserve before fill" approach provides better
+       behavior than "fill then notify" in presence of reentrant
+       modifications (which may occur below, on a poll point in the loop or
+       the [Elem] allocation):
+
+       - If some code asynchronously adds new elements after this
+         length update, they will go after the space we just reserved,
+         and in particular no addition will be lost. If instead we
+         updated the length after the loop, any asynchronous addition
+         during the loop could be erased or erase one of our additions,
+         silently, without warning the user.
+
+       - If some code asynchronously iterates on the dynarray, or
+         removes elements, or otherwise tries to access the
+         reserved-but-not-yet-filled space, it will get a clean "missing
+         element" error. This is worse than with the fill-then-notify
+         approach where the new elements would only become visible
+         (to iterators, for removal, etc.) alltogether at the end of
+         loop.
+
+       To summarise, "reserve before fill" is better on add-add races,
+       and "fill then notify" is better on add-remove or add-iterate
+       races. But the key difference is the failure mode:
+       reserve-before fails on add-remove or add-iterate races with
+       a clean error, while notify-after fails on add-add races with
+       silently disappearing data. *)
+    for i = 0 to length_b - 1 do
+      let x = Array.unsafe_get b i in
+      Array.unsafe_set arr (length_a + i) (Elem {v = x})
+    done;
+    true
+  end
+
+let append_array a b =
+  if append_array_if_room a b then ()
+  else begin
+    (* slow path *)
+    let rec grow_and_append a b =
+      ensure_extra_capacity a (Array.length b);
+      if not (append_array_if_room a b)
+      then grow_and_append a b
+    in grow_and_append a b  end
+
+(* append: same [..._if_room] and loop logic as [add_last],
+   same reserve-before-fill logic as [append_array]. *)
+
+(* Note: unlike [add_last_if_room], [append_if_room] is *not* atomic.
+
+   It is a programming error to mutate the length of [b] during a call
+   to [append a b]. To detect this mistake we keep track of the length
+   of [b] throughout the computation and check it that does not
+   change.
+*)
+let append_if_room a b ~length_b =
+  (* BEGIN ATOMIC *)
+  let {arr = arr_a; length = length_a} = a in
+  if length_a + length_b > Array.length arr_a then false
+  else begin
+    a.length <- length_a + length_b;
+    (* END ATOMIC *)
+    let arr_b = b.arr in
+    check_valid_length length_b arr_b;
```

Comment:
(For context: the function only fails if the length of `b` is detected, which is ruled out by the specification as a programming error as the function iterates on `b`. So we are talking about an invalid state arising in the case of an exception that is not intended to be recovered from. The current implementation is specification-conformant and reasonable in practice but @damiendoligez wonder if we can do even slightly better by avoiding the inconsistent state.)

One aspect of this function that I think is important is that the correct-usage case should go as fast as reasonably possible -- modulo the requirement of detecting invalidation and reporting them propeerly. The only way I see to avoid creating inconsistent states (assuming we consider only the single-domain / reentrant scenario) is to never reserve space in `a` for several consecutive elements but add them one by one. I worry that this would result in a noticeable performance overhead in the loop, with an extra write to the length field. (Maybe this is absorbed by the allocation cost, but I am skeptical, and that would stop being true in an unboxed version.)